src/EditorFeatures/Core/Implementation/AddImports/AbstractAddImportsPasteCommandHandler.cs

```diff
@@ -0,0 +1,105 @@
+
+        private readonly IThreadingContext _threadingContext;
+
+        public AbstractAddImportsPasteCommandHandler(IThreadingContext threadingContext)
+            => _threadingContext = threadingContext;
+
+        public CommandState GetCommandState(PasteCommandArgs args, Func<CommandState> nextCommandHandler)
+            => nextCommandHandler();
+
+        public void ExecuteCommand(PasteCommandArgs args, Action nextCommandHandler, CommandExecutionContext executionContext)
+        {
+            // Capture the pre-paste caret position
+            var caretPosition = args.TextView.GetCaretPoint(args.SubjectBuffer);
+            if (!caretPosition.HasValue)
+            {
+                return;
+            }
+
+            // Create a tracking span from the pre-paste caret position that will grow as text is inserted.
+            var trackingSpan = caretPosition.Value.Snapshot.CreateTrackingSpan(caretPosition.Value.Position, 0, SpanTrackingMode.EdgeInclusive);
+
+            // Perform the paste command before adding imports
+            nextCommandHandler();
+
+            if (!args.SubjectBuffer.CanApplyChangeDocumentToWorkspace())
+            {
+                return;
+            }
+
+            // Don't perform work if we're inside the interactive window
+            if (args.TextView.IsBufferInInteractiveWindow(args.SubjectBuffer))
+            {
+                return;
+            }
+
+            if (!args.SubjectBuffer.GetFeatureOnOffOption(FeatureOnOffOptions.AddImportsOnPaste))
+            {
+                return;
+            }
+
+            // Applying the post-paste snapshot to the tracking span gives us the span of pasted text.
+            var snapshotSpan = trackingSpan.GetSpan(args.SubjectBuffer.CurrentSnapshot);
+            var textSpan = snapshotSpan.Span.ToTextSpan();
+
+            var sourceTextContainer = args.SubjectBuffer.AsTextContainer();
+            if (!Workspace.TryGetWorkspace(sourceTextContainer, out var workspace))
+            {
+                return;
+            }
+
+            var document = sourceTextContainer.GetOpenDocumentInCurrentContext();
+            if (document is null)
+            {
+                return;
+            }
+
+            var addMissingImportsService = document.GetLanguageService<IAddMissingImportsFeatureService>();
+            if (addMissingImportsService is null)
+            {
+                return;
+            }
+
+            using var _ = executionContext.OperationContext.AddScope(allowCancellation: true, DialogText);
+            var cancellationToken = executionContext.OperationContext.UserCancellationToken;
+
+#pragma warning disable VSTHRD102 // Implement internal logic asynchronously
+            var updatedDocument = _threadingContext.JoinableTaskFactory.Run(() => addMissingImportsService.AddMissingImportsAsync(document, textSpan, cancellationToken));
+#pragma warning restore VSTHRD102 // Implement internal logic asynchronously
```

Comment:
is this what we do elsewhere?  I'm not super familiar with .OperationContext.  but if this is our regular pattern then 👍

Reply:
OperationContext just requires that we have a disposable scope, which will block the UI interaction. Since command handlers are synchronous, this is the only way to take the async work and do it all in a synchronous fashion. If we go full async, we no longer will need an OperationContext or the disposable scope.